lib/mainloop.c

```diff
@@ -297,6 +298,27 @@ main_loop_reload_config_initiate(gpointer user_data)
   main_loop_worker_sync_call(main_loop_reload_config_apply, self);
 }
 
+static void
+block_till_workers_exit()
+{
+  gint64 end_time;
```

Comment:
are the worker threads signaled to exit at this point? e.g. IIRC iv_work_pool_put() should wake up the workers and cause them to exit in a timely manner.

our threaded destination code might be blocking for some time still, but I think that's ok.

Reply:
> iv_work_pool_put

I call this block after main_loop_io_worker_deinit(), which calls iv_work_pool_put.
  main_loop_io_worker_deinit();                                                                                                        
  main_loop_worker_deinit();                                                                                                           
  block_till_workers_exit(); 

> our threaded destination code might be blocking for some time still, but I think that's ok.

For the threaded destinations: currently the complete exit mechanism waits for them: the exit mechanism is executed with main_loop_worker_sync_call, which waits for all threaded destinations to call main_loop_worker_job_complete(). So yes, the threaded destination drivers can block the exit, but the block happens earlier than block_till_workers_exit() called, so the timeout with g_cond_wait_until is not affected.